cibersort/cibersort.xml

```diff
@@ -0,0 +1,170 @@
+        expression data.
+    </description>
+
+    <macros>
+        <import>macros.xml</import>
+    </macros>
+
+    <expand macro="requirements" />
+
+    @VERSION_CMD@
+
+    <command detect_errors="exit_code"><![CDATA[
+        Rscript $run
+    ]]></command>
+
+    <configfiles>
+        <configfile name="run">
+source(paste0("${__tool_directory__}","/CIBERSORT.R"))
+library(dplyr)
+library(reshape2)
+library(ggplot2)
+library(RColorBrewer)
+
+# Define the function used to produce the plots
+graph = function(data, level, colors) {
+  if (level == "CellType") {
+    writeName = "Raw"
+  } else {
+    writeName = level
+  }
+
+  ggplot(data, aes_string(x=level, y="EstimatedPercent", fill="Sample")) +
+    geom_bar(stat="identity", position='dodge') +
+    theme_minimal() +
+    theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
+    scale_fill_brewer(palette="Set2") 
+  filename = paste(c("CIBERSORT",writeName,"barplot.png"),collapse="_")
+  ggsave(filename)
+  
+  ggplot(data,aes_string(x="Sample", y="EstimatedPercent", fill=level)) +
+    geom_bar(stat = "identity", position = "stack") +
+    theme_minimal() +
+    theme(axis.text.x = element_text(angle = 90)) +
+    scale_fill_manual(values = colors) +
+    guides(fill=guide_legend(ncol=1))
+  filename = paste(c("CIBERSORT",writeName,"stacked.png"),collapse="_")
+  ggsave(filename)
+}
+
+# Define the function used to determine color scheme
+getColors = function(cellTypes, map=NA, level=NA) {
+  levelColor = paste0(level,"Color")
+  if ("$mapping.provided"=="FALSE" || toupper("$mapping.map_colors")==F 
+      || is.na(map) || is.na(level) || !(levelColor %in% colnames(map))
+      || is.na(map[,levelColor])) {
+    colorCount = length(cellTypes)
+    getPalette = colorRampPalette(brewer.pal(9, "Set1"))
+    colors = rev(getPalette(colorCount))
+  } else {
+    colors = setNames(as.character(map[,levelColor]), map[,level])
+  }
+  return(colors)
+}
+
+# Define the function that extracts an array of valid column titles for the bulk
+# file
+convertBulkColNames = function(samples) {
+    asEntered = unlist(strsplit(samples, ","))
+    subDots = lapply(asEntered,FUN=function(x){return(gsub('-','\\.',x))})
+    addX = lapply(subDots,FUN=function(x){return(paste0('X',x))})
+    return(unlist(addX))
+}
+
+# Set up required data
+signature = read.csv("$signature_file", header=TRUE, row.names=1, sep="\t")
+bulk = as.matrix(read.csv("$expression_file", header=TRUE, row.names=1, sep="\t"))
+samples = convertBulkColNames("$sample_ids")
+##bulk = as.matrix(select(bulk, unlist(samples)))
+bulk = as.matrix(bulk)
+
+res = CIBERSORT(signature, bulk,QN=TRUE, absolute=FALSE, abs_method="sig.score")
+write.table(res, "$output_tsv", quote=FALSE, sep="\t")
+
+cols = ncol(signature)
+cells = res[,1:cols] *100
+cells = cells[samples,,drop=FALSE]
+data = melt(cells)
+colnames(data) = c("Sample", "CellType", "EstimatedPercent")
+
+#if $mapping.provided == "FALSE"
```

Comment:
Switch to cheetah/python style boolean instead of R?

Reply:
It's not _really_ a boolean, since it used as the select in a conditional parameter. To the best of my knowledge, you can't use real booleans in that context. The choice to use `"TRUE"` use `"FALSE"` is mostly because it results in more natural expression in the script, in my opinion. Can be replaced with `"provided"` and `"omitted"` if preferred.

Reply:
It seems like this logic could be handled by either cheetah or R. IMO, it would be more intuitive if it is not written to appear as if combining both. One way to do that is by changing the value as you point out. Other ideas that could be explored:
```suggestion
if ($mapping.provided == "FALSE") {
```
or 
```suggestion
#if $mapping.map_file:
```